app/scripts/controllers/metamask-notifications/metamask-notifications.ts

```diff
@@ -0,0 +1,954 @@
+      return this.messagingSystem.call('AuthenticationController:isSignedIn');
+    },
+  };
+
+  #storage = {
+    getStorageKey: () => {
+      return this.messagingSystem.call('UserStorageController:getStorageKey');
+    },
+    getNotificationStorage: async () => {
+      return await this.messagingSystem.call(
+        'UserStorageController:performGetStorage',
+        'notification_settings',
+      );
+    },
+    setNotificationStorage: async (state: string) => {
+      return await this.messagingSystem.call(
+        'UserStorageController:performSetStorage',
+        'notification_settings',
+        state,
+      );
+    },
+  };
+
+  #pushNotifications = {
+    enablePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:enablePushNotifications',
+        UUIDs,
+      );
+    },
+    disablePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:disablePushNotifications',
+        UUIDs,
+      );
+    },
+    updatePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:updateTriggerPushNotifications',
+        UUIDs,
+      );
+    },
+  };
+
+  #accountsCache = new Set<string>();
+
+  #accounts = {
+    listAccounts: () => {
+      // Fetch the list of accounts
+      const accounts = this.messagingSystem.call(
+        'AccountsController:listAccounts',
+      ) as InternalAccount[];
+
+      // Extract addresses and convert them to checksum addresses to ensure case sensitivity is not an issue
+      const addresses = accounts.map((account) => {
+        return toChecksumHexAddress(account.address as string);
+      });
+
+      this.#accountsCache = new Set([...this.#accountsCache, ...addresses]);
+      return [...this.#accountsCache];
+    },
+
+    // This ensures that our cache is up-to-date when constructor is called.
+    initialize: () => {
+      this.#accounts.listAccounts();
+    },
+
+    // Subscription when an new account is changed
+    // The initial cache should already have the addresses in wallet,
+    // so any new account switched to (that is not in cache) We can assume it is new.
+    createNewNotificationsOnNewAccountChange: () => {
+      this.messagingSystem.subscribe(
+        'AccountsController:selectedAccountChange',
+        async ({ address }: { address: string }) => {
+          if (!this.state.isMetamaskNotificationsEnabled || !address) {
+            return;
+          }
+
+          const checksumAddress = toChecksumHexAddress(address);
+          if (this.#accountsCache.has(checksumAddress)) {
+            return;
+          }
+
+          // An actual new address! We can create triggers/notifications for this.
+          this.#accountsCache.add(checksumAddress);
+          await this.updateOnChainTriggersByAccount(checksumAddress);
```

Comment:
Does this also need to be called for the accounts cached in the `listAccounts` method?

Reply:
This has now been covered, we do make sure that we support multi-add or remove accounts.